src/masternodeman.cpp

```diff
@@ -1308,11 +1308,18 @@ bool CMasternodeMan::CheckMnbAndUpdateMasternodeList(CMasternodeBroadcast mnb, i
     nDos = 0;
     LogPrint("masternode", "CMasternodeMan::CheckMnbAndUpdateMasternodeList -- masternode=%s\n", mnb.vin.prevout.ToStringShort());
 
-    if(mapSeenMasternodeBroadcast.count(mnb.GetHash())) { //seen
+    uint256 hash = mnb.GetHash();
+    if(mapSeenMasternodeBroadcast.count(hash)) { //seen
         LogPrint("masternode", "CMasternodeMan::CheckMnbAndUpdateMasternodeList -- masternode=%s seen\n", mnb.vin.prevout.ToStringShort());
+        // less then 2 pings left before this MN goes into non-recoverable state, bump sync timeout
+        if(GetTime() - mapSeenMasternodeBroadcast[hash].first > MASTERNODE_NEW_START_REQUIRED_SECONDS - MASTERNODE_MIN_MNP_SECONDS * 2) {
+            LogPrint("masternode", "CMasternodeMan::CheckMnbAndUpdateMasternodeList -- masternode=%s seen update\n", mnb.vin.prevout.ToStringShort());
+            mapSeenMasternodeBroadcast[hash].first = GetTime();
```

Comment:
I'm not 100% sure about that one either tbh. Primarily the reason behind this is the question "ok, so we requested all mnb from our peers but did we actually receive anything?". In 12.0 that was kind of easy to answer this question because if you were offline for more than expiry time (~1h), mns were removed and list was resynced from scratch which was bumping `nTimeLastMasternodeList`. In 12.1 however we keep the list, so we almost never bump it now even if we receive some mnb because most of them are already in `seen` map. So I guessed that sync could time out and switch to next asset too fast sometimes and keeping the time mnb was received we'll be able to bump it again which could help in dangerous situations like switching mns in non-recoverable state. So basically, this just restores the old behavior but probably won't help _that_ much in complex situations if initial sync stales for some reason.